sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java

```diff
@@ -674,6 +680,85 @@ public String toString() {
       abstract Builder<T> setClock(@Nullable Clock clock);
 
       abstract Read<T> build();
+
+      @Override
+      public PTransform<PBegin, PCollection<T>> buildExternal(External.Configuration config) {
+        if (config.topic != null) {
+          StaticValueProvider<String> topic = StaticValueProvider.of(utf8String(config.topic));
+          setTopicProvider(NestedValueProvider.of(topic, new TopicTranslator()));
+        }
+        if (config.subscription != null) {
+          StaticValueProvider<String> subscription =
+              StaticValueProvider.of(utf8String(config.subscription));
+          setSubscriptionProvider(
+              NestedValueProvider.of(subscription, new SubscriptionTranslator()));
+        }
+        if (config.idAttribute != null) {
+          String idAttribute = utf8String(config.idAttribute);
+          setIdAttribute(idAttribute);
+        }
+        if (config.timestampAttribute != null) {
+          String timestampAttribute = utf8String(config.timestampAttribute);
+          setTimestampAttribute(timestampAttribute);
+        }
+        setNeedsAttributes(config.needsAttributes);
+        setPubsubClientFactory(FACTORY);
+        if (config.needsAttributes) {
+          SimpleFunction<PubsubMessage, T> parseFn =
+              (SimpleFunction<PubsubMessage, T>) new IdentityMessageFn();
+          setParseFn(parseFn);
+          // FIXME: call setCoder(). need to use PubsubMessage proto to be compatible with python
```

Comment:
In the case that an external sdk has requested `needsAttributes`, this transform needs to produce `PubsubMessage` instances rather than byte arrays.  There's a wrinkle here:  python deserializes `PubsubMessage` using protobufs, whereas Java encodes with `PubsubMessageWithAttributesCoder`. 

I believe we want to use protobufs from Java.  Can I get confirmation of that, please?

Reply:
I serialized the `PubsubMessage` using protobufs.  Since there's no cross-language coder for `PubsubMessage`, and I assumed it would be overreach to add one, I used the bytes coder and then handled converting to and from protobufs in code that lives close to the transforms.